File: LeetCode_CPP/num034/num034.cpp
// File: num034.cpp
// Author: lxw
// Date: 2015-06-13

/*
Num num034: Search for a Range
Source: https://leetcode.com/problems/search-for-a-range/
Given a sorted array of integers, find the starting and ending position of a given target value.

Your algorithm's runtime complexity must be in the order of O(log n).

If the target is not found in the array, return [-1, -1].

For example,
Given [5, 7, 7, 8, 8, 10] and target value 8,
return [3, 4].
*/

class Solution {
public:
    int bsConvert(vector<int>& nums, int begin, int end, int target){
        int middle = 0;
        while(begin <= end){
            middle = (begin + end) / 2;
            if(nums[middle] == target){
                return middle;
            }
            else if(nums[middle] > target){
                end = middle - 1;
            }
            else{
                begin = middle + 1;
            }
        }
        return -1;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        int length = nums.size();
        vector<int> vec(2, -1);
        if(length == 0){
            return vec;
        }
        int index = bsConvert(nums, 0, length-1, target);
        if(index == -1){
            return vec;
        }
        int i = index;
        int j = index;
        while(nums[i] == target && i >= 0){
            --i;
        }
        ++i;
        while(nums[j] == target && j < length){
            ++j;
        }
        --j;
        vec[0] = i;
        vec[1] = j;
        return vec;
    }
};

File: LeetCode_Python/num034/num034.py
#!/usr/bin/python2.7
# FileName: num071.py
# Author: lxw
# Date: 2015-07-03

"""
Num 034: 
Source: 

Given a sorted array of integers, find the starting and ending position of a given target value.

Your algorithm's runtime complexity must be in the order of O(log n).

If the target is not found in the array, return [-1, -1].

For example,
Given [5, 7, 7, 8, 8, 10] and target value 8,
return [3, 4].
"""

class Solution(object):
    def binSearch(self, nums, target, start, end):
        while start <= end:
            mid = (start + end) >> 1
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                end = mid - 1
            else:
                start = mid + 1
        return -1

    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        length = len(nums)
        index = self.binSearch(nums, target, 0, length-1)
        if index == -1:
            return [-1, -1]
        else:
            start = index
            end = index
            while start >= 0 and nums[start] == target:
                start -= 1
            start += 1

            while end < length and nums[end] == target:
                end += 1
            end -= 1
            return [start, end]

if __name__ == '__main__':
    sol = Solution()
    print sol.searchRange([1, 3, 3, 4, 5, 5, 5, 7, 7, 8, 8], 5)

